examples/spatial.pct.py

```diff
@@ -0,0 +1,259 @@
+ax.set(title="Switzerland's topography and SwissMetNet stations", aspect="auto")
+cb = fig.colorbar(ax.collections[-1], aspect=50, pad=0.02)
+cb.set_label("Max. daily temperature [°C]", labelpad=-2)
+
+
+# %% [markdown]
+# As always, we store our training data in a `Dataset` object.
+# %%
+from gpjax.dataset import Dataset
+
+x = temperature[["latitude", "longitude", "elevation"]].values
+y = temperature[["t_max"]].values
+D = Dataset(
+    X=jnp.array(x),
+    y=jnp.array(y),
+)
+
+# %% [markdown]
+# ## ARD Kernel
+# As temperature decreases with height
+# (at a rate of approximately -6.5 °C/km in average conditions), we can expect that using the geographical distance
+# alone isn't enough to to a decent job at interpolating this data. Therefore, we can also use elevation and optimize
+# the parameters of our kernel such that more relevance should be given to elevation. This is possible by using a
+# kernel that has one length-scale parameter per input dimension: an automatic relevance determination (ARD) kernel.
+
+# %%
+kernel = gpx.kernels.RBF(
+    active_dims=[0, 1, 2],
+    lengthscale=jnp.array([0.1, 0.1, 100.0]),
+)
+
+# %% [markdown]
+# ## Mean function
+# As stated before, we already know that temperature strongly depends on elevation.
+# So why not use it for our mean function? GPJax lets you define custom mean functions
+# easily: simply subclass `AbstractMeanFunction`.
+
+# %%
+from dataclasses import dataclass
+from jaxtyping import Float, Array
+from gpjax.base import param_field
+
+
+@dataclass
+class MeanFunction(gpx.gps.AbstractMeanFunction):
+    w: Float[Array, "1"] = param_field(jnp.array([0.0]))
+    b: Float[Array, "1"] = param_field(jnp.array([0.0]))
+
+    def __call__(self, x: Float[Array, "N D"]) -> Float[Array, "N 1"]:
+        elevation = x[:, 2:3]
+        out = elevation * self.w + self.b
+        return out
+
+
+# %% [markdown]
+# Now we can define our prior. We'll also choose a Gaussian likelihood.
+
+# %%
+mean_function = MeanFunction()
+prior = gpx.Prior(kernel, mean_function)
+likelihood = gpx.Gaussian(D.n)
+
+# %% [markdown]
+# Finally we construct the posterior.
+# %%
+posterior = prior * likelihood
+print(posterior)
+
+
+# %% [markdown]
+# ## Model fitting
+# We proceed to train our model. Because we used a Gaussian likelihood, the resulting posterior is
+# a `ConjugatePosterior`, which allows us to optimize the analytically expressed marginal loglikelihood.
+#
+# As always, we can jit-compile the objective function to speed things up.
+# %%
+
+negative_mll = jax.jit(gpx.objectives.ConjugateMLL(negative=True))
+negative_mll(posterior, train_data=D)
+
+# %%
+import optax as ox
+
+optim = ox.chain(ox.adam(learning_rate=0.1), ox.clip(1.0))
+posterior, history = gpx.fit(
+    model=posterior,
+    objective=negative_mll,
+    train_data=D,
+    optim=optim,
+    num_iters=3000,
+    safe=True,
+)
+
+posterior: gpx.gps.ConjugatePosterior
```

Comment:
What is this line doing?

Reply:
Yeah I was going to bring this up btw. The fit method accepts and returns `Module` types and IDE language servers (e.g. pylance) won't know that `posterior` is of type `ConjugatePosterior` in this case. So no "mouse hovering" for automatic docstring display, no tab completion, no methods suggestions. By adding this line we tell the language server the type. I can remove it for now but would be nice to think of a way (overloading?) to return the same specific type that was passed to `fit`. Just a user-friendliness thing.

Reply:
Neat. I didn't know this was possible. Let's discuss it further on Thursday but, for now, this line makes sense.